### PR TITLE
Fixed problem with adding existing wallet for the second time.

### LINKED ISSUE #85: Adding this same wallet address in read mode cause error 500
1. Log in to the app
2. Go to the wallets
3. Add read only wallet with name `mywallet1` and address `z1x2...c3`
4. Add second wallet with different name (`mywallet2`) and this same address `z1x2...c3`

Observe 500 Internal Server Error (that is persistent on `/wallets` route).

### PR DESCRIPTION
**Pull request by furmanp**
Fix for the following: 
https://github.com/emeth-defi-automation/portfolio-management-platform/issues/85

### CODE REVIEW

File: src/utils/validators/addWallet.ts [hunk 1]
```diff
@@ -58,6 +58,32 @@ export const isNameUnique = server$(async function (name: string) {
...
+  }
+  const [usersObservedWallets] = UniqueAddressResult.parse(result);
+  for (const observedWalletAddress of usersObservedWallets) {
+    if (observedWalletAddress === address) {
+      return false;
+    }
+  }
+  return true;
+});
+
```
Comment (Uacias) [inline src/utils/validators/addWallet.ts]: I believe we could simplify this query to:
`SELECT count() FROM wallet WHERE address = type::string(address)"
This should work, lmk if im wrong

**Review by Uacias — CHANGES_REQUESTED** [review]
Overall looks fine, I'd only request query modification, IMO it could be simplified.

File: src/utils/validators/addWallet.ts (hunk 1 above)
Comment (furmanp) [reply src/utils/validators/addWallet.ts]: updated - made it slightly different though. Your query is not user scoped.

File: src/utils/validators/addWallet.ts [hunk 2]
```diff
@@ -58,6 +58,29 @@ export const isNameUnique = server$(async function (name: string) {
...
+  const cookie = this.cookie.get("accessToken");
+  if (!cookie) {
+    throw new Error("No cookie found");
+  }
+  const { userId } = jwt.decode(cookie.value) as JwtPayload;
+  const result = (
+    await db.query(`
+     LET $walletId = SELECT VALUE ->observes_wallet.out FROM ${userId};
+     LET $wallets = SELECT address FROM wallet WHERE $walletId;
+     SELECT VALUE count() FROM $wallets WHERE address = type::string('${address}');
```
Comment (Uacias) [inline src/utils/validators/addWallet.ts:73]: Could we aspire to shorten the query to look more similiar to:
```
const balanceToUpdate = (
    await db.query(
      `SELECT * FROM balance WHERE ->(for_wallet WHERE out.address = '${getAddress(accountAddress)}') AND ->(for_token WHERE out.address = '${getAddress(tokenAddress)}');`,
    )
  ).at(0);
  ```

**Review by Uacias — CHANGES_REQUESTED** [review]
Overall good job, altho I'd suggest shortening the query. 
If my approach is unoptimal lmk.

File: src/utils/validators/addWallet.ts (hunk 2 above)
Comment (furmanp) [reply src/utils/validators/addWallet.ts:73]: putting it here for the record.
Following your proposal the query would look as follows:
```
SELECT VALUE count() FROM wallet WHERE (
    SELECT ->observes_wallet.out FROM "${user}"
) AND address = type::string('${address}');
```

I run a quick benchmark and it seems that despite having longer query, using LET's improves code performance.
Current query: 0.3us on average
Modified one (one above):  1.1ms on average

Comment (Uacias) [reply src/utils/validators/addWallet.ts:73]: Great research!
We will stay with your approach in that case.

**Review by Uacias — APPROVED** [review]
Microwave

**Pull request merged**